8280761: UseCompressedOops should be set after limit_heap_by_allocatable_memory

tkiriyama opened the pull request:
I fixed to set UseCompressedOops flag after limit_heap_by_allocatable_memory(). 
So when ulimit -v is called and -XX:MaxRAM is set, UseCompressedOops does not become false. 
And all hotspot tier1 test are passed.
Would you please review this fix?

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8280761](https://bugs.openjdk.java.net/browse/JDK-8280761): UseCompressedOops should be set after limit_heap_by_allocatable_memory


### Reviewers
 * [Albert Mingkun Yang](https://openjdk.java.net/census#ayang) (@albertnetymk - **Reviewer**) ⚠️ Review applies to e2fe253f33e8937ba334ff1e3218f86d0ac6fda0
 * [Thomas Schatzl](https://openjdk.java.net/census#tschatzl) (@tschatzl - **Reviewer**) ⚠️ Review applies to 148d0fd6454108d1a7f13a4c11df5c219cc62b49


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7938/head:pull/7938` \
`$ git checkout pull/7938`

Update a local copy of the PR: \
`$ git checkout pull/7938` \
`$ git pull https://git.openjdk.java.net/jdk pull/7938/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7938`

View PR using the GUI difftool: \
`$ git pr show -t 7938`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7938.diff">https://git.openjdk.java.net/jdk/pull/7938.diff</a>

</details>

tkiriyama commented:
Could anyone review this pull request, please?

dholmes-ora commented:
/cc hotspot-gc

dholmes-ora commented:
I cc'd hotspot-gc as they should probably be the ones to look at this for you.

tkiriyama commented:
Hello, I would appreciate it if anyone could review my fix.

albertnetymk commented:
I plan to take a look this week. Sorry for the delay; I was occupied by some other reviewing work last week.

albertnetymk commented:
I noticed that Github Action is not enabled for this PR. Please follow the instructions in https://wiki.openjdk.java.net/display/SKARA/Testing to enable Github Action. Also, I got build errors locally.

tkiriyama commented:
> I noticed that Github Action is not enabled for this PR. Please follow the instructions in https://wiki.openjdk.java.net/display/SKARA/Testing to enable Github Action. Also, I got build errors locally.

I am very sorry. I enabled Github Action .
I have successfully built on Windows and Linux locally.

albertnetymk commented:
> I enabled Github Action .

Doesn't look like GA is properly running. It should say sth like "12 successful checks", as shown in #7341.

> I have successfully built on Windows and Linux locally.

OK, maybe this is a 32/64-bit thing. Having a working GA would have caught this.

tschatzl commented:
The code will not compile on anything but 32 bit platforms. Here's a result on linux-x64:

```
.../src/hotspot/share/runtime/arguments.cpp: In static member function 'static void Arguments::set_heap_size()':
.../src/hotspot/share/runtime/arguments.cpp:1826:11: error: 'reasonable_max' was not declared in this scope
 1826 |       if (reasonable_max > max_coop_heap) {
```

`reasonable_max` used within the `_LP64` block, and the change ends the scope it is declared in at line 1795. This also results in the entire block in the `_LP64` block to be misaligned. I believe (without checking), the closing brace at 1795 has been intended to stay in line 1842.

Please fix.

albertnetymk commented on test/hotspot/jtreg/gc/arguments/TestUseCompressedOopsFlagsWithUlimit.java:
```diff
@@ -0,0 +1,95 @@
+
+  public static void main(String args[]) throws Exception {
+    // Tests
+    // 1. Verify that MaxRAMPercentage overrides UseCompressedOops Ergo
+    // 2. Verify that UseCompressedOops forces compressed oops limit even
+    //    when other flags are specified.
```
I don't understand why `MaxRAMPercentage` or `UseCompressedOops` are relevant to this fix/bug.

albertnetymk commented on test/hotspot/jtreg/gc/arguments/TestUseCompressedOopsFlagsWithUlimit.java:
```diff
@@ -0,0 +1,95 @@
+    args.add("-version");
+
+    String cmd = ProcessTools.getCommandLine(ProcessTools.createTestJvm(args.toArray(String[]::new)));
+    ProcessBuilder pb = new ProcessBuilder("sh", "-c", "ulimit -v 10485760;" + cmd);
```
The `ulimit -v` parameter should be passed in as well like `MaxRAM`. Then, from the calling context, one can see whether coop should be enabled or not.

tkiriyama replied:
Thank you for your review.

> I don't understand why `MaxRAMPercentage` or `UseCompressedOops` are relevant to this fix/bug.

I forgot to remove this comment when I copied form TestMaxRAMFlags.java.
I fixed this comment.

> The `ulimit -v` parameter should be passed in as well like `MaxRAM`. Then, from the calling context, one can see whether coop should be enabled or not.

This bug appears when ulimit -v is specified and MaxRAM is set bigger value. So I think it does not need that
ulimit -v parameter let be variable.

albertnetymk replied:
It's not about whether it's constant or variable.

In this context, it's unclear why coop is enabled.

```java
// Args: MaxRAM , MaxRAMPercentage, forcecoop, expect coop
checkFlag(32 * oneG, 100, false, true);
```

`checkFlag` should not sneak in logic altering the coop value.

albertnetymk reviewed (changes requested):
The hotspot change looks good to me, but there's sth unclear to me in the test.

albertnetymk commented on test/hotspot/jtreg/gc/arguments/TestUseCompressedOopsFlagsWithUlimit.java:
```diff
@@ -0,0 +1,95 @@
+  public static void main(String args[]) throws Exception {
+    // Tests
+    // 1. Verify that UseCompressedOops Ergo follows ulimit -v setting.
+    // 2. Verify that UseCompressedOops forces compressed oops limit even
+    //    when ulimit -v are specified.
```
I understand the first bullet but not the second. Looking at the reproducers in the JBS ticket, `UseCompressedOops` should not be provided; it's read only.

tschatzl replied:
This looks like an additional (unrelated) test to make sure it is overridden if necessary imho. It seems a bit forced though because there is no test that actually verifies that `UseCompressedOops` is `false` later.
I'm good with removing the test case.

tkiriyama replied:
> This looks like an additional (unrelated) test to make sure it is overridden if necessary imho.

That is correct.  
Do you mean that the following code should be removed?  

`checkFlag(128 * oneG, 100, true, true);`

tkiriyama replied:
I understand.  
I removed this test patern.

tschatzl reviewed (approved):
Looks okay.

albertnetymk commented on test/hotspot/jtreg/gc/arguments/TestUseCompressedOopsFlagsWithUlimit.java:
```diff
@@ -0,0 +1,95 @@
+
+    long oneG = 1L * 1024L * 1024L * 1024L;
+
+    // Args: MaxRAM , MaxRAMPercentage, forcecoop, expect coop
```
I am not sure why `MaxRAMPercentage` is needed here.

tschatzl replied:
I think this is that the initial value for `reasonable_max` is exactly `MaxRAM`; you are right that for this test it is not necessary to pass it explicitly, just always set it to 100 percent in the command line.

albertnetymk replied:
Aha, I see. Having `MaxRAMPercentage` set explicitly instead of relying on its default value indeed makes the test more resilient. Please add a comment for `MaxRAMPercentage`.

tkiriyama replied:
OK.  
I added a comment for MaxRAMPercentage.

albertnetymk reviewed (commented):
In summary, the signature I have in mind is sth like `checkFlag(long ulimit_v, long max_ram, boolean expected_coop)`.

albertnetymk commented on test/hotspot/jtreg/gc/arguments/TestUseCompressedOopsFlagsWithUlimit.java:
```diff
@@ -0,0 +1,89 @@
+    long oneG = 1L * 1024L * 1024L * 1024L;
+
+    // Args: MaxRAM , MaxRAMPercentage, expect coop
+    // Having MaxRAMPercentage set explicitly instead of relying on its default value indeed makes the test more resilient.
```
Slight rewording, `// Setting MaxRAMPercentage explicitly to make the test more resilient.`

tkiriyama commented:
> Slight rewording, // Setting MaxRAMPercentage explicitly to make the test more resilient.

Thank you, I fixed this comment.

> Embedding ulimit -v inside checkFlag still hinders readability, IMO.

Sounds good.  I fixed this tests as you advised.

albertnetymk commented:
Need sth like this to resolve the failure.

```java
  // Convert bytes to kbytes for ulimit -v
  var ulimit_prefix = "ulimit -v " + (ulimit / 1024);
  ...
  ProcessBuilder pb = new ProcessBuilder("sh", "-c", ulimit_prefix + ";" + cmd);
```

tkiriyama commented:
> Need sth like this to resolve the failure.

I'm sorry. This is my mistake. I added your code.

tkiriyama commented:
I appreciate all reviews. I hope this change is integrated.

tkiriyama commented:
/integrate

albertnetymk commented:
/sponsor
